File: app/src/main/java/club/dev/mobile/ksu/takeanumber/Firebase/Repository.java
package club.dev.mobile.ksu.takeanumber.Firebase;

import android.arch.core.util.Function;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Transformations;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.util.ArrayList;
import java.util.List;

import club.dev.mobile.ksu.takeanumber.Data.HelpSession;
import club.dev.mobile.ksu.takeanumber.Data.Student;

public class Repository {
    private FirebaseDatabase mDatabase;
    private static Repository mSoleInstance;
    private DatabaseReference mReference;
    private LiveData<List<HelpSession>> mHelpSessions;

    public static Repository getInstance() {
        if (mSoleInstance == null) {
            mSoleInstance = new Repository();
        }
        return mSoleInstance;
    }

    private Repository() {
        mDatabase = FirebaseDatabase.getInstance();
        mReference = mDatabase.getReference();
        mHelpSessions = Transformations.map(
                new HelpSessionLiveData(mReference.child("HelpSession2")),
                new HelpSessionDeserializer());
    }

    public void addHelpSession(HelpSession session) {
        String key = mReference.child("HelpSession2").push().getKey();
        session.setFirebaseKey(key);
        mReference.child("HelpSession2").child(key).setValue(session);
    }

    public void addStudentToQueue(Student student, String sessionKey) {
        DatabaseReference ref = mReference.child("HelpSession2").child(sessionKey).child("queue");
        String key = ref.push().getKey();
        student.setFirebaseKey(key);
        ref.child(key).setValue(student);
    }

    public LiveData<List<Student>> getStudentQueue(String sessionKey) {
        return Transformations.map(new StudentQueueLiveData(
                mReference.child("HelpSession2").child(sessionKey).child("queue").orderByChild("dateTime")),
                new StudentQueueDeserializer()
        );
    }

    public void deleteStudent(String sessionName, String studentKey)
    {
        mReference.child("HelpSession2").child(sessionName).child("queue").child(studentKey).removeValue();
    }

    public void updateStudent(String sessionName, Student student)
    {
        mReference.child("HelpSession2").child(sessionName).child("queue").child(student.getFirebaseKey()).setValue(student);
    }

    public LiveData<List<HelpSession>> getHelpSessions() {
        return mHelpSessions;
    }

    private class HelpSessionDeserializer implements Function<DataSnapshot, List<HelpSession>> {

        @Override
        public List<HelpSession> apply(DataSnapshot input) {
            List<HelpSession> sessionList = new ArrayList<>();
            for (DataSnapshot snap : input.getChildren()) {
                HelpSession session = snap.getValue(HelpSession.class);
                session.setFirebaseKey(snap.getKey());
                sessionList.add(session);
            }
            return sessionList;
        }
    }

    private class StudentQueueDeserializer implements Function<DataSnapshot, List<Student>> {

        @Override
        public List<Student> apply(DataSnapshot input) {
            List<Student> students = new ArrayList<>();
            for (DataSnapshot snap : input.getChildren()) {
                Student student = snap.getValue(Student.class);
                student.setFirebaseKey(snap.getKey());
                students.add(student);
            }
            return students;
        }
    }
}

File: app/src/main/java/club/dev/mobile/ksu/takeanumber/Firebase/Repository.kt
package club.dev.mobile.ksu.takeanumber.Firebase

import androidx.arch.core.util.Function
import androidx.lifecycle.LiveData
import androidx.lifecycle.Transformations
import club.dev.mobile.ksu.takeanumber.Data.HelpSession
import club.dev.mobile.ksu.takeanumber.Data.Student
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import java.util.*

class Repository private constructor() {

    private val mDatabase: FirebaseDatabase = FirebaseDatabase.getInstance()
    private val mReference: DatabaseReference = mDatabase.reference
    val helpSessions: LiveData<List<HelpSession>>

    fun addHelpSession(session: HelpSession) {
        val key = mReference.child("HelpSession").push().key
        session.firebaseKey = key
        mReference.child("HelpSession").child(key!!).setValue(session)
    }

    fun addStudentToQueue(student: Student, sessionKey: String) {
        val ref = mReference.child("HelpSession").child(sessionKey).child("queue")
        val key = ref.push().key
        student.firebaseKey = key
        ref.child(key!!).setValue(student)
    }

    fun getStudentQueue(sessionKey: String): LiveData<List<Student>> {
        return Transformations.map<DataSnapshot, List<Student>>(
            StudentQueueLiveData(
                mReference.child("HelpSession").child(sessionKey).child("queue")
                    .orderByChild("dateTime")
            ),
            StudentQueueDeserializer()
        )
    }

    fun deleteStudent(sessionName: String, studentKey: String) {
        mReference.child("HelpSession").child(sessionName).child("queue").child(studentKey)
            .removeValue()
    }

    fun updateStudent(sessionName: String, student: Student) {
        mReference.child("HelpSession").child(sessionName).child("queue")
            .child(student.firebaseKey).setValue(student)
    }

    private inner class HelpSessionDeserializer : Function<DataSnapshot, List<HelpSession>> {
        override fun apply(input: DataSnapshot): List<HelpSession> {
            val sessionList: MutableList<HelpSession> = ArrayList()
            for (snap in input.children) {
                val session = snap.getValue(HelpSession::class.java)!!
                session.firebaseKey = snap.key
                sessionList.add(session)
            }
            return sessionList
        }
    }

    private inner class StudentQueueDeserializer : Function<DataSnapshot, List<Student>> {
        override fun apply(input: DataSnapshot): List<Student> {
            val students: MutableList<Student> = ArrayList()
            for (snap in input.children) {
                val student = snap.getValue(Student::class.java)!!
                student.firebaseKey = snap.key
                students.add(student)
            }
            return students
        }
    }

    companion object {
        private var mSoleInstance: Repository? = null
        @JvmStatic
        val instance: Repository?
            get() {
                if (mSoleInstance == null) {
                    mSoleInstance = Repository()
                }
                return mSoleInstance
            }
    }

    init {
        helpSessions = Transformations.map<DataSnapshot, List<HelpSession>>(
            HelpSessionLiveData(mReference.child("HelpSession")),
            HelpSessionDeserializer()
        )
    }
}